sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumns.scala

```diff
@@ -15,139 +15,385 @@
-   * had performed an exhaustive backfill of the provided value to all previously existing rows in
-   * the table instead. We choose to avoid doing such a backfill because it would be a
-   * time-consuming and costly operation. Instead, we elect to store the EXISTS_DEFAULT in the
-   * column metadata for future reference when querying data out of the data source. In turn, each
-   * data source then takes responsibility to provide the constant-folded value in the
-   * EXISTS_DEFAULT metadata for such columns where the value is not present in storage.
-   *
-   * @param analyzer used for analyzing the result of parsing the column expression stored as text.
-   * @param tableSchema represents the names and types of the columns of the statement to process.
-   * @param statementType name of the statement being processed, such as INSERT; useful for errors.
-   * @return a copy of `tableSchema` with field metadata updated with the constant-folded values.
+   * @param analyzer analyzer to use for processing DEFAULT values stored as text.
+   * @param catalog  the catalog to use for looking up the schema of INSERT INTO table objects.
+   * @param insert   the enclosing INSERT statement for which this rule is processing the query, if
+   *                 any.
    */
-  def constantFoldCurrentDefaultsToExistDefaults(
+  case class ResolveDefaultColumns(
       analyzer: Analyzer,
-      tableSchema: StructType,
-      statementType: String): StructType = {
-    if (!SQLConf.get.enableDefaultColumns) {
-      return tableSchema
+      catalog: SessionCatalog,
+      insert: Option[InsertIntoStatement] = None) extends Rule[LogicalPlan] {
+    override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
+      (_ => SQLConf.get.enableDefaultColumns), ruleId) {
+      case i@InsertIntoStatement(_, _, _, _, _, _)
+        if i.query.collectFirst { case u: UnresolvedInlineTable => u }.isDefined =>
+        // Create a helper instance of this same rule with the `insert` argument populated as `i`.
+        // Then recursively apply it on the `query` of `i` to transform the result. This recursive
+        // application lets the below case matching against `UnresolvedInlineTable` to trigger
+        // anywhere that operator may appear in the descendants of `i.query`.
+        val helper = ResolveDefaultColumns(analyzer, catalog, Some(i))
+        val newQuery = helper.apply(i.query)
+        i.copy(query = newQuery)
+
+      case table: UnresolvedInlineTable
+        if insert.isDefined &&
+          table.rows.nonEmpty && table.rows.forall(_.size == table.rows(0).size) =>
+        val expanded: UnresolvedInlineTable = addMissingDefaultColumnValues(table).getOrElse(table)
+        replaceExplicitDefaultColumnValues(analyzer, expanded).getOrElse(table)
+
+      case i@InsertIntoStatement(_, _, _, project: Project, _, _) =>
+        val helper = ResolveDefaultColumns(analyzer, catalog, Some(i))
```

Comment:
why do we still need to recursively create another `ResolveDefaultColumns`

Reply:
This was just to populate the `insert` field in the case class. Since changing this to a `var`, we can do without the recursion now.